File: sdk/python/cds/sources/companies.py
"""
SignedData CDS — Companies Brazil Fetcher
Source: BrasilAPI — https://brasilapi.com.br/api/cnpj/v1/{cnpj}

Query-driven: no scheduled ingestor. Called per CNPJ lookup from the MCP server.
"""

from __future__ import annotations

import hashlib
import re
from datetime import UTC, datetime
from typing import Any

import httpx

from cds.schema import CDSEvent, ContextMeta, SourceMeta
from cds.sources.companies_models import (
    CNAECode,
    CompaniesContentTypes,
    CompanyAddress,
    CompanyPartner,
    CompanyPartners,
    CompanyProfile,
)
from cds.vocab import CDSSources

BRASILAPI_CNPJ_BASE = "https://brasilapi.com.br/api/cnpj/v1"
BRASILAPI_CNAE_BASE = "https://brasilapi.com.br/api/cnae/v1"


def _fingerprint(data: bytes) -> str:
    return "sha256:" + hashlib.sha256(data).hexdigest()


def _format_cnpj(bare: str) -> str:
    """Format bare CNPJ digits: '11222333000144' → '11.222.333/0001-44'"""
    return f"{bare[:2]}.{bare[2:5]}.{bare[5:8]}/{bare[8:12]}-{bare[12:14]}"


def _brl_capital(value: float) -> str:
    """Format BRL: 205431960490.52 → 'R$ 205.431.960.490,52'"""
    formatted = f"{value:,.2f}".replace(",", "X").replace(".", ",").replace("X", ".")
    return f"R$ {formatted}"


def validate_cnpj(cnpj: str) -> str:
    """
    Validate CNPJ check digits and return normalised bare digits.
    Supports both '11.222.333/0001-44' and '11222333000144'.
    Raises ValueError on invalid CNPJ.
    """
    bare = re.sub(r"[.\-/]", "", cnpj)
    if len(bare) != 14 or not bare.isdigit():
        raise ValueError(f"CNPJ must have 14 digits, got: {cnpj!r}")

    # All same digits is invalid
    if len(set(bare)) == 1:
        raise ValueError(f"Invalid CNPJ (all same digits): {cnpj!r}")

    # First check digit
    weights1 = [5, 4, 3, 2, 9, 8, 7, 6, 5, 4, 3, 2]
    total = sum(int(bare[i]) * weights1[i] for i in range(12))
    remainder = total % 11
    d1 = 0 if remainder < 2 else 11 - remainder
    if int(bare[12]) != d1:
        raise ValueError(f"Invalid CNPJ check digit 1: {cnpj!r}")

    # Second check digit
    weights2 = [6, 5, 4, 3, 2, 9, 8, 7, 6, 5, 4, 3, 2]
    total = sum(int(bare[i]) * weights2[i] for i in range(13))
    remainder = total % 11
    d2 = 0 if remainder < 2 else 11 - remainder
    if int(bare[13]) != d2:
        raise ValueError(f"Invalid CNPJ check digit 2: {cnpj!r}")

    return bare


def _parse_profile(raw: dict[str, Any], query_ts: str) -> CompanyProfile:
    """Parse BrasilAPI CNPJ response into CompanyProfile."""
    cnpj_bare = str(raw.get("cnpj", "")).zfill(14)
    main_cnae_code = str(raw.get("cnae_fiscal", "")).zfill(7)
    main_cnae_desc = raw.get("cnae_fiscal_descricao", "")

    secondary = []
    for s in raw.get("cnaes_secundarios", []):
        code = str(s.get("codigo", "")).zfill(7)
        desc = s.get("descricao", "")
        if code != "0000000":
            secondary.append(CNAECode(code=code, description=desc))

    address = CompanyAddress(
        street=raw.get("logradouro", ""),
        number=raw.get("numero", ""),
        complement=raw.get("complemento") or None,
        neighborhood=raw.get("bairro", ""),
        zip_code=str(raw.get("cep", "")),
        city=raw.get("municipio", ""),
        state=raw.get("uf", ""),
    )

    return CompanyProfile(
        cnpj=cnpj_bare,
        cnpj_formatted=_format_cnpj(cnpj_bare),
        company_name=raw.get("razao_social", ""),
        trade_name=raw.get("nome_fantasia") or None,
        registration_status=raw.get("descricao_situacao_cadastral", "ATIVA"),
        registration_date=raw.get("data_inicio_atividade", ""),
        registration_status_date=raw.get("data_situacao_cadastral"),
        legal_nature_code=str(raw.get("codigo_natureza_juridica", "")),
        legal_nature=raw.get("natureza_juridica", ""),
        size=raw.get("porte", "DEMAIS"),
        share_capital=raw.get("capital_social"),
        main_cnae=CNAECode(code=main_cnae_code, description=main_cnae_desc),
        secondary_cnaes=secondary,
        address=address,
        phone=raw.get("ddd_telefone_1") or None,
        email=raw.get("email") or None,
        query_timestamp=query_ts,
    )


def _parse_partners(raw: dict[str, Any], query_ts: str) -> CompanyPartners:
    """Parse BrasilAPI CNPJ response QSA into CompanyPartners."""
    cnpj_bare = str(raw.get("cnpj", "")).zfill(14)
    partners = []
    for qsa in raw.get("qsa", []):
        partners.append(
            CompanyPartner(
                name=qsa.get("nome_socio", ""),
                qualifier=qsa.get("qualificacao_socio", ""),
                qualifier_code=int(qsa.get("codigo_qualificacao_socio", 0)),
                entry_date=qsa.get("data_entrada_sociedade"),
                country=qsa.get("pais"),
                legal_representative=qsa.get("nome_representante_legal") or None,
            )
        )
    return CompanyPartners(
        cnpj=cnpj_bare,
        company_name=raw.get("razao_social", ""),
        partners=partners,
        query_timestamp=query_ts,
    )


def _build_profile_summary(profile: CompanyProfile) -> str:
    """Build a human-readable summary for a company profile."""
    name = profile.trade_name or profile.company_name
    status = profile.registration_status
    parts = [f"{name} ({profile.cnpj_formatted}): {status}"]

    if profile.registration_date:
        parts[0] += f" desde {profile.registration_date}"

    parts.append(f"CNAE: {profile.main_cnae.description}")

    if profile.share_capital:
        parts.append(f"Capital social: {_brl_capital(profile.share_capital)}")

    return ". ".join(parts) + "."


class CNPJFetcher:
    """
    Fetches and signs CNPJ data on demand.
    Not a scheduled ingestor — called per query from the MCP server.
    """

    def __init__(self, signer: Any = None) -> None:
        self.signer = signer

    async def fetch_profile(self, cnpj: str) -> CDSEvent:
        """Fetch full company profile. Validates CNPJ format first."""
        bare = validate_cnpj(cnpj)
        query_ts = datetime.now(UTC).isoformat()

        async with httpx.AsyncClient(timeout=15) as client:
            resp = await client.get(f"{BRASILAPI_CNPJ_BASE}/{bare}")
            resp.raise_for_status()
            fp = _fingerprint(resp.content)
            raw = resp.json()

        profile = _parse_profile(raw, query_ts)
        event = CDSEvent(
            content_type=CompaniesContentTypes.PROFILE,
            source=SourceMeta(id=CDSSources.BRASILAPI, fingerprint=fp),
            occurred_at=datetime.now(UTC),
            lang="pt-BR",
            payload=profile.model_dump(mode="json"),
            event_context=ContextMeta(
                summary=_build_profile_summary(profile),
                model="rule-based-v1",
            ),
        )
        if self.signer:
            self.signer.sign(event)
        return event

    async def fetch_partners(self, cnpj: str) -> CDSEvent:
        """Fetch partner/shareholder data (QSA)."""
        bare = validate_cnpj(cnpj)
        query_ts = datetime.now(UTC).isoformat()

        async with httpx.AsyncClient(timeout=15) as client:
            resp = await client.get(f"{BRASILAPI_CNPJ_BASE}/{bare}")
            resp.raise_for_status()
            fp = _fingerprint(resp.content)
            raw = resp.json()

        partners = _parse_partners(raw, query_ts)
        n = len(partners.partners)
        event = CDSEvent(
            content_type=CompaniesContentTypes.PARTNERS,
            source=SourceMeta(id=CDSSources.BRASILAPI, fingerprint=fp),
            occurred_at=datetime.now(UTC),
            lang="pt-BR",
            payload=partners.model_dump(mode="json"),
            event_context=ContextMeta(
                summary=f"{partners.company_name}: {n} sócio{'s' if n != 1 else ''}",
                model="rule-based-v1",
            ),
        )
        if self.signer:
            self.signer.sign(event)
        return event

File: sdk/typescript/src/sources/companies.ts
/**
 * SignedData CDS — Companies Brazil Source
 * Source: BrasilAPI — https://brasilapi.com.br/api/cnpj/v1/{cnpj}
 *
 * Query-driven: no scheduled ingestor. Called per CNPJ lookup.
 */

import { createHash } from "node:crypto";
import { CDSEvent, type ContextMeta, SourceMeta } from "../schema.js";
import type { CDSSigner } from "../signer.js";
import { CDSVocab, CDSSources } from "../vocab.js";

// ── Content Types ──────────────────────────────────────────

export const CompaniesContentTypes = {
  PROFILE:  CDSVocab.COMPANIES_PROFILE_CNPJ,
  PARTNERS: CDSVocab.COMPANIES_PARTNERS_CNPJ,
  CNAE:     CDSVocab.COMPANIES_CNAE_PROFILE,
} as const;

// ── Payload Types ──────────────────────────────────────────

export interface CNAECode {
  code: string;
  description: string;
}

export interface CompanyAddress {
  street: string;
  number: string;
  complement?: string;
  neighborhood: string;
  zip_code: string;
  city: string;
  state: string;
}

export interface CompanyProfile {
  cnpj: string;
  cnpj_formatted: string;
  company_name: string;
  trade_name?: string;
  registration_status: "ATIVA" | "BAIXADA" | "INAPTA" | "SUSPENSA" | "NULA";
  registration_date: string;
  registration_status_date?: string;
  legal_nature_code: string;
  legal_nature: string;
  size: string;
  share_capital?: number;
  main_cnae: CNAECode;
  secondary_cnaes: CNAECode[];
  address: CompanyAddress;
  phone?: string;
  email?: string;
  query_timestamp: string;
}

export interface CompanyPartner {
  name: string;
  qualifier: string;
  qualifier_code: number;
  entry_date?: string;
  country?: string;
  legal_representative?: string;
}

export interface CompanyPartners {
  cnpj: string;
  company_name: string;
  partners: CompanyPartner[];
  query_timestamp: string;
}

// ── CNPJ Validation ────────────────────────────────────────

export function validateCnpj(cnpj: string): string {
  const bare = cnpj.replace(/[.\-/]/g, "");
  if (bare.length !== 14 || !/^\d+$/.test(bare)) {
    throw new Error(`CNPJ must have 14 digits, got: ${cnpj}`);
  }
  if (new Set(bare).size === 1) {
    throw new Error(`Invalid CNPJ (all same digits): ${cnpj}`);
  }

  const weights1 = [5, 4, 3, 2, 9, 8, 7, 6, 5, 4, 3, 2];
  let total = 0;
  for (let i = 0; i < 12; i++) total += Number(bare[i]) * weights1[i];
  let remainder = total % 11;
  const d1 = remainder < 2 ? 0 : 11 - remainder;
  if (Number(bare[12]) !== d1) {
    throw new Error(`Invalid CNPJ check digit 1: ${cnpj}`);
  }

  const weights2 = [6, 5, 4, 3, 2, 9, 8, 7, 6, 5, 4, 3, 2];
  total = 0;
  for (let i = 0; i < 13; i++) total += Number(bare[i]) * weights2[i];
  remainder = total % 11;
  const d2 = remainder < 2 ? 0 : 11 - remainder;
  if (Number(bare[13]) !== d2) {
    throw new Error(`Invalid CNPJ check digit 2: ${cnpj}`);
  }

  return bare;
}

export function formatCnpj(bare: string): string {
  return `${bare.slice(0, 2)}.${bare.slice(2, 5)}.${bare.slice(5, 8)}/${bare.slice(8, 12)}-${bare.slice(12, 14)}`;
}

// ── Helpers ────────────────────────────────────────────────

const BRASILAPI_CNPJ_BASE = "https://brasilapi.com.br/api/cnpj/v1";

type R = Record<string, unknown>;

function fingerprint(data: Buffer): string {
  return "sha256:" + createHash("sha256").update(data).digest("hex");
}

// ── Fetcher ────────────────────────────────────────────────

export class CNPJFetcher {
  constructor(private readonly signer?: CDSSigner) {}

  async fetchProfile(cnpj: string): Promise<CDSEvent> {
    const bare = validateCnpj(cnpj);
    const queryTs = new Date().toISOString();

    const resp = await globalThis.fetch(`${BRASILAPI_CNPJ_BASE}/${bare}`);
    if (!resp.ok) throw new Error(`BrasilAPI HTTP ${resp.status}`);
    const buf = Buffer.from(await resp.arrayBuffer());
    const fp = fingerprint(buf);
    const raw = JSON.parse(buf.toString("utf-8")) as R;

    const cnpjBare = String(raw["cnpj"] ?? "").padStart(14, "0");
    const profile: CompanyProfile = {
      cnpj: cnpjBare,
      cnpj_formatted: formatCnpj(cnpjBare),
      company_name: (raw["razao_social"] as string) ?? "",
      trade_name: (raw["nome_fantasia"] as string) || undefined,
      registration_status: (raw["descricao_situacao_cadastral"] as CompanyProfile["registration_status"]) ?? "ATIVA",
      registration_date: (raw["data_inicio_atividade"] as string) ?? "",
      legal_nature_code: String(raw["codigo_natureza_juridica"] ?? ""),
      legal_nature: (raw["natureza_juridica"] as string) ?? "",
      size: (raw["porte"] as string) ?? "DEMAIS",
      share_capital: raw["capital_social"] as number | undefined,
      main_cnae: {
        code: String(raw["cnae_fiscal"] ?? "").padStart(7, "0"),
        description: (raw["cnae_fiscal_descricao"] as string) ?? "",
      },
      secondary_cnaes: ((raw["cnaes_secundarios"] as R[]) ?? [])
        .map(s => ({
          code: String(s["codigo"] ?? "").padStart(7, "0"),
          description: (s["descricao"] as string) ?? "",
        }))
        .filter(s => s.code !== "0000000"),
      address: {
        street: (raw["logradouro"] as string) ?? "",
        number: (raw["numero"] as string) ?? "",
        complement: (raw["complemento"] as string) || undefined,
        neighborhood: (raw["bairro"] as string) ?? "",
        zip_code: String(raw["cep"] ?? ""),
        city: (raw["municipio"] as string) ?? "",
        state: (raw["uf"] as string) ?? "",
      },
      phone: (raw["ddd_telefone_1"] as string) || undefined,
      email: (raw["email"] as string) || undefined,
      query_timestamp: queryTs,
    };

    const name = profile.trade_name ?? profile.company_name;
    const summary = `${name} (${profile.cnpj_formatted}): ${profile.registration_status}`;

    const event = new CDSEvent({
      content_type: CompaniesContentTypes.PROFILE,
      source: { "@id": CDSSources.BRASILAPI, fingerprint: fp },
      occurred_at: queryTs,
      lang: "pt-BR",
      payload: profile as unknown as R,
      context: {
        summary,
        model: "rule-based-v1",
        generated_at: queryTs,
      } satisfies ContextMeta,
    });

    if (this.signer) this.signer.sign(event);
    return event;
  }

  async fetchPartners(cnpj: string): Promise<CDSEvent> {
    const bare = validateCnpj(cnpj);
    const queryTs = new Date().toISOString();

    const resp = await globalThis.fetch(`${BRASILAPI_CNPJ_BASE}/${bare}`);
    if (!resp.ok) throw new Error(`BrasilAPI HTTP ${resp.status}`);
    const buf = Buffer.from(await resp.arrayBuffer());
    const fp = fingerprint(buf);
    const raw = JSON.parse(buf.toString("utf-8")) as R;

    const partners: CompanyPartners = {
      cnpj: String(raw["cnpj"] ?? "").padStart(14, "0"),
      company_name: (raw["razao_social"] as string) ?? "",
      partners: ((raw["qsa"] as R[]) ?? []).map(q => ({
        name: (q["nome_socio"] as string) ?? "",
        qualifier: (q["qualificacao_socio"] as string) ?? "",
        qualifier_code: Number(q["codigo_qualificacao_socio"] ?? 0),
        entry_date: (q["data_entrada_sociedade"] as string) || undefined,
        country: (q["pais"] as string) || undefined,
        legal_representative: (q["nome_representante_legal"] as string) || undefined,
      })),
      query_timestamp: queryTs,
    };

    const n = partners.partners.length;
    const event = new CDSEvent({
      content_type: CompaniesContentTypes.PARTNERS,
      source: { "@id": CDSSources.BRASILAPI, fingerprint: fp },
      occurred_at: queryTs,
      lang: "pt-BR",
      payload: partners as unknown as R,
      context: {
        summary: `${partners.company_name}: ${n} sócio${n !== 1 ? "s" : ""}`,
        model: "rule-based-v1",
        generated_at: queryTs,
      } satisfies ContextMeta,
    });

    if (this.signer) this.signer.sign(event);
    return event;
  }
}
